### PR TITLE
[SPARK-6483][SQL]Improve ScalaUdf called performance.

### PR DESCRIPTION
**Pull request by zzcclp**
As issue [SPARK-6483](https://issues.apache.org/jira/browse/SPARK-6483) description, ScalaUdf is low performance because of calling _asInstanceOf_ to convert per record.
With this, the performance of ScalaUdf is the same as other case.
thank @lianhuiwang for telling me how to resolve this problem.

### CODE REVIEW

**Comment by chenghao-intel** [comment]
Hmm, have you try what the performance gain by this change? From my understanding the bottleneck is in the function call `ScalaReflection.convertToScala`

**Comment by zzcclp** [comment]
Before this change, it takes 17 minutes, and now takes 5 minutes, which is the same as *HiveContext + udf floor* and *non-udf*

**Comment by chenghao-intel** [comment]
OK, probably we can also move the `children.size match {..}` out of the `eval`.

**Comment by chenghao-intel** [comment]
I mean we can do something like
```scala
val f = children.size match {
  case 1 =>
   val func = function.asInstanceOf[(Any) => Any]
   val child0 = children(0)
    (input: Row) => {
      func(ScalaReflection.convertToScala(child0.eval(input), child0.dataType)))
    }

  case 2 =>
   val func = function.asInstanceOf[(Any) => Any]
   val child0 = children(0)
   val child1 = children(1)

   (input: Row) => {
    func(ScalaReflection.convertToScala(child0.eval(input), child0.dataType))
       ScalaReflection.convertToScala(child1.eval(input), child1.dataType)))
    }
}

def eval(input: Row) = f(input)
```

**Comment by zzcclp** [comment]
OK, I will modify code and test again.

**Comment by zzcclp** [comment]
@chenghao-intel , I change code and test it, the result is the same as last commit  , is 5 minutes.
Please help me for checking code.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUdf.scala [hunk 1]
```diff
@@ -49,353 +49,378 @@ case class ScalaUdf(function: AnyRef, dataType: DataType, children: Seq[Expressi
...
+            ScalaReflection.convertToScala(children(12).eval(input), children(12).dataType),
+            ScalaReflection.convertToScala(children(13).eval(input), children(13).dataType),
+            ScalaReflection.convertToScala(children(14).eval(input), children(14).dataType),
+            ScalaReflection.convertToScala(children(15).eval(input), children(15).dataType),
+            ScalaReflection.convertToScala(children(16).eval(input), children(16).dataType),
+            ScalaReflection.convertToScala(children(17).eval(input), children(17).dataType),
+            ScalaReflection.convertToScala(children(18).eval(input), children(18).dataType),
+            ScalaReflection.convertToScala(children(19).eval(input), children(19).dataType),
+            ScalaReflection.convertToScala(children(20).eval(input), children(20).dataType),
+            ScalaReflection.convertToScala(children(21).eval(input), children(21).dataType))
```
Comment (chenghao-intel) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUdf.scala]: `children` is the type of `Seq[Expression]` (essentially the `List[Expression]`), access its element by index will cause performance overhead in runtime, we'd better move that out of the anonymous functions. See:
http://docs.scala-lang.org/overviews/collections/performance-characteristics.html

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUdf.scala [hunk 2]
```diff
@@ -49,353 +49,378 @@ case class ScalaUdf(function: AnyRef, dataType: DataType, children: Seq[Expressi
...
-          ScalaReflection.convertToScala(children(20).eval(input), children(20).dataType),
-          ScalaReflection.convertToScala(children(21).eval(input), children(21).dataType))
-
-    }
-    // scalastyle:on
-
-    ScalaReflection.convertToCatalyst(result, dataType)
+  
+  val f = children.size match {
+      case 0 =>
```
Comment (chenghao-intel) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUdf.scala]: This code block can be generated by the scala script, as we in the master branch, can you also update/paste the code gen script?
The original code can be found at:
https://github.com/zzcclp/spark/blob/SPARK-6483/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUdf.scala#L38

**Comment by chenghao-intel** [comment]
@zzcclp I will run the benchmark in my local machine, will get back soon.

@liancheng , can you trigger the unit test?

**Comment by chenghao-intel** [comment]
Verified the code change by the following micro-benchmark
```scala
import org.apache.spark.sql.catalyst.expressions._
import org.apache.spark.sql.types._

case class Floor(child: Expression) extends UnaryExpression with Predicate {
  override def foldable = child.foldable
  def nullable = child.nullable
  override def toString = s"Floor $child"

  override def eval(input: Row): Any = {
    child.eval(input) match {
      case null => null
      case ts: Int => ts - ts % 300
    }
  }
}

object T {
  def benchmark(count: Int, expr: Expression): Unit = {
    var i = 0
    val row = new GenericRow(Array[Any](123, 21, 42))
    val s = System.currentTimeMillis()
    while (i < count) {
      expr.eval(row)
      i += 1
    }
    val e = System.currentTimeMillis()

    println (s"${expr.getClass.getSimpleName}  -- ${e - s} ms")
  }
  def main(args: Array[String]) {
    def func(ts: Int) = ts - ts % 300
    val udf0 = ScalaUdf(func _, IntegerType, BoundReference(0, IntegerType, true) :: Nil)
    val udf1 = Floor(BoundReference(0, IntegerType, true))

    benchmark(1000000, udf0)
    benchmark(1000000, udf0)
    benchmark(1000000, udf0)

    benchmark(1000000, udf1)
    benchmark(1000000, udf1)
    benchmark(1000000, udf1)
  }
}
```

Without the code change it outputs
ScalaUdf  -- 1183 ms
ScalaUdf  -- 887 ms
ScalaUdf  -- 929 ms

Floor  -- 49 ms
Floor  -- 15 ms
Floor  -- 21 ms


With the code change, it outputs
ScalaUdf  -- 382 ms
ScalaUdf  -- 255 ms
ScalaUdf  -- 247 ms

Floor  -- 27 ms
Floor  -- 6 ms
Floor  -- 8 ms

Conclusions: 
* The code change will improve the performance of scala udf by 2-3x
* Scala UDF is in very low performance compare to the built-in type of Expression.

We probably need to provide more efficient way of UDF extension interface.

**Comment by liancheng** [comment]
ok to test

**Comment by zzcclp** [comment]
@chenghao-intel ,I have update the code. can you take a look again. thanks.

**Comment by chenghao-intel** [comment]
You need to fetch the latest code and resolve the conflicts.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUdf.scala [hunk 3]
```diff
@@ -30,372 +30,423 @@ case class ScalaUdf(function: AnyRef, dataType: DataType, children: Seq[Expressi
   type EvaluatedType = Any
 
   def nullable = true
-
```
Comment (chenghao-intel) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUdf.scala]: Nit: revert the change here?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUdf.scala [hunk 4]
```diff
@@ -30,372 +30,423 @@ case class ScalaUdf(function: AnyRef, dataType: DataType, children: Seq[Expressi
...
   // scalastyle:off
+  
+  override def toString = s"scalaUDF(${children.mkString(",")})"
 
   /** This method has been generated by this script
 
-    (1 to 22).map { x =>
-      val anys = (1 to x).map(x => "Any").reduce(_ + ", " + _)
-      val evals = (0 to x - 1).map(x => s"    ScalaReflection.convertToScala(children($x).eval(input), children($x).dataType)").reduce(_ + ",\n    " + _)
+        (1 to 22).map { x =>
```
Comment (chenghao-intel) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUdf.scala]: Nit: indention issues, we should keep 2 spaces instead of 4.

**Comment by chenghao-intel** [comment]
@zzcclp LGTM, except some small issues.

**Comment by zzcclp** [comment]
@SparkQA , merge again.

**Comment by zzcclp** [comment]
@chenghao-intel , I have update the code. can you take a look again. thanks.

@AmplabJenkins , re-test

**Comment by chenghao-intel** [comment]
@zzcclp 
children:Seq[Expression] essentially can be either `List[Expression]` or `ArraySeq[Expression]`, the later doesn't support the pattern matching.

Probably you have to use the code like:
```scala
val f = children.size match {
  case 1 =>
   val func = function.asInstanceOf[(Any) => Any]
   val child0 = children(0)
    (input: Row) => {
      func(ScalaReflection.convertToScala(child0.eval(input), child0.dataType)))
    }

  case 2 =>
   val func = function.asInstanceOf[(Any) => Any]
   val child0 = children(0)
   val child1 = children(1)

   (input: Row) => {
    func(ScalaReflection.convertToScala(child0.eval(input), child0.dataType))
       ScalaReflection.convertToScala(child1.eval(input), child1.dataType)))
    }
}
```

**Comment by zzcclp** [comment]
@chenghao-intel , can you review again, thanks.

**Comment by zzcclp** [comment]
@AmplabJenkins , please re test

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUdf.scala [hunk 5]
```diff
@@ -39,363 +39,668 @@ case class ScalaUdf(function: AnyRef, dataType: DataType, children: Seq[Expressi
...
+          ScalaReflection.convertToScala(child20.eval(input), child20.dataType),
+          ScalaReflection.convertToScala(child21.eval(input), child21.dataType))
+      }
   }
-}
+  
+  // scalastyle:on
+  
+  override def eval(input: Row): Any = ScalaReflection.convertToCatalyst(f(input), dataType)
+}
```
Comment (chenghao-intel) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUdf.scala]: A newline should be added at the end of source file.

**Comment by zzcclp** [comment]
Add a newline at the end of source file, does it need to re test?

**Comment by zzcclp** [comment]
@AmplabJenkins , please re test, thanks.

**Comment by chenghao-intel** [comment]
The unit test will be auto-triggered, once the code changed, you needn't say anything to @AmplabJenkins .

**Comment by zzcclp** [comment]
OK, I am new sparker, :smile:

**Comment by liancheng** [comment]
@zzcclp @chenghao-intel Thanks for working on this and the review comments! Merged to master.

**Comment by liancheng** [comment]
@zzcclp Would you please set your real name on both GitHub and JIRA so that our script can put your name on the credit list of the next release? Also, it would be good if you can set your name in git config:

```
$ git config --global user.name "Your Name"
```

**Comment by zzcclp** [comment]
@liancheng , I already set my real name on my GitHub *Name* and JIRA *Full Name*. :smiley:
